src/provider/controller/AsyncInit.ts

```diff
@@ -0,0 +1,24 @@
+import {injectable} from "inversify";
+
+/**
+ * Base class for any objects that initialise asynchronously.
+ * 
+ * Since constructors cannot be awaited, this class acts to create a common pattern for these types of objects.
+ * 
+ * To ensure the object is fully initilased before usage, place an `await myObject.initialised;` immediately after 
+ * construction.
+ */
+@injectable()
+export abstract class AsyncInit {
+    private _initialized: Promise<this>;
+
+    constructor() {
+        this._initialized = this.init().then(() => this);
+    }
+
+    public get initialized(): Promise<this> {
```

Comment:
Yeah. Though in reality, not likely to ever be needed.

The `await` of these `AsyncInit` classes is now done within the `Injector` util (though this happens after all objects have been created, rather than pre-injection), so _I think_ we shouldn't need to do an `await foo.init()` anywhere in the "main" provider code.